File: haskell/src/Tree/BinaryTree.hs
module Tree.BinaryTree
( EmptyTree
, BinaryTree (..)
, singleton
, insert
, delete
) where

type EmptyTree = None

data BinaryTree a = EmptyTree | Node a { value :: a, left :: (BinaryTree a), right :: (BinaryTree a) } deriving (Eq, Show, Read)

insert :: (Ord a) => a 

File: go/src/tree/binary_tree.go
package tree

type BinaryTree struct {
  left *BinaryTree
  right *BinaryTree
  value interface{}
}

// preorder walk of the binary tree
func (b *BinaryTree) Walk() []interface{} {
  var left, right []interface{}

  if b.left != nil {
    left = b.left.Walk()
  }

  if b.right != nil {
    right = b.right.Walk()
  }

  return append(append(left, b.value), right...)
}
